udev: make OPTIONS="string_escape=replace" take effect on ENV{key}= assiginment

yuwata opened the pull request:
Fixes #19309.

mwhudson commented:
Hi, this does seem to fix the problem, thanks! I wasn't able to figure out how to make qemu set the model to something with a slash in, but it's easy to put / into the serial which produced the same issue, which was resolved by this patch.

poettering commented on man/udev.xml:
```diff
@@ -605,7 +605,7 @@
                 </varlistentry>
                 <varlistentry>
-                  <term><option>string_escape=<replaceable>none|replace</replaceable></option></term>
+                  <term><option>string_escape=<replaceable>none|replace|slash</replaceable></option></term>
```
this should explain what this does.

poettering commented on src/udev/udev-rules.c:
```diff
@@ -2012,6 +2018,19 @@ static int udev_rule_apply_token_to_event(
+                        if (count > 0)
+                                log_rule_debug(dev, rules, "Replaced %zu shash(es) from result of ENV{%s}%s=\"%s\"",
+                                               count, name, token->op == OP_ADD ? "+" : "", token->value);
+                }
```
maybe make this a function of its own?

poettering commented:
Hmm, so the other ID_SERIAL values (i.e. for usb and such) all are careful in escaping all weird chars with udev_replace_chars()

Only for nvme we don't do this, because it's done in a udev rule instead of helper. I figure the right fix for this would be to use udev_replace_chars() here too, i.e. escape these serials the same way as all others.

This would probably mean we'd have to add this kind of escaping to rules, too, at leas optional.

/cc @msekletar who added the original NVME patch for this in a5110c90303cf455db5062faef34d5724d12e2e9

I wonder how much the mangling for this would affect the serial strings IRL. At least on my system the serials don#t have such weird chars, so I guess in many cases we'll just be fine.

poettering commented:
long story short: i thin the approach is OK, but we should go further and process the strings with the full power of udev_replace_chars() instead of a lightweight escaper that only processes "/".

yuwata commented:
@poettering Thank you for your review, and sorry for late response. Now, `string_escape=replace` option is reused for `ENV{}=` assignment. As I commented in the commit message, strictly speaking, this breaks backward compatibility. PTAL.

poettering commented on src/udev/udev-rules.c:
```diff
@@ -2022,6 +2022,12 @@ static int udev_rule_apply_token_to_event(
                         l = strpcpyl(&p, l, val, " ", NULL);
 
                 (void) udev_event_apply_format(event, token->value, p, l, false);
+                if (event->esc == ESCAPE_REPLACE) {
```
hmm, if this is done for ENV{} assignments too now, I fiugre the man page explanation for string_escape= should be slightly updated.

it currently says:

>  Usually, control and other possibly unsafe characters are replaced in strings used for device naming. The mode of replacement can be specified with this option.

ENV{} doesn't really fit under the "strings used for device naming" umbrella. 

also, it should probably say that for the device naming stuff we default to escaping but for the env stuff we default to not escaping.

poettering commented:
hmm, thinking about this more: can this actually work? i.e. wouldn't the ENV{} assignment ecape once and then the SYMLINKS= stuff escape a second time? so we'd end up with a doubly-escaped symlink?

yuwata commented:
Updated. Now a new netif naming scheme is also added. PTAL.

> hmm, thinking about this more: can this actually work? i.e. wouldn't the ENV{} assignment ecape once and then the SYMLINKS= stuff escape a second time? so we'd end up with a doubly-escaped symlink?

@poettering Why double escaping do you think causes problem here? `string_escape=replace` just replaces characters, not 'real' escaping, so the operation is idempotent.

keszybz commented on src/udev/udev-rules.c:
```diff
@@ -2053,21 +2053,20 @@ static int udev_rule_apply_token_to_event(
 
+                if (sd_device_get_ifindex(dev, NULL) < 0) {
+                        log_rule_error(dev, rules,
+                                       "Only network interface can be renamed, ignoring NAME=\"%s\"; please fix it.",
```
"please fix it" is unclear: fix what and how? Also, the fact that this is printed in red is enough to tell the reader that it's something to fix.

> `"Only network interfaces can be renamed, ignoring NAME=\"%s\"."`

keszybz reviewed (approved):
LGTM.

The log message can always be adjusted after -rc2, the behaviour change is what is important.
